compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/transformers/body/resolve/FirBodyResolveTransformer.kt

```diff
@@ -64,11 +65,21 @@ open class FirBodyResolveTransformer(
         val resolvedTypeRef = if (typeRef is FirResolvedTypeRef) {
             typeRef
         } else {
+
             typeResolverTransformer.withFile(context.file) {
+
+                val selfAnnotation =
+                    context.topClassDeclaration?.annotations?.find { it.typeRef.render() == "R|kotlin/Self|" }
+                val scopes =
+                    components.createCurrentScopeList() + if (context.topClassDeclaration != null && selfAnnotation != null)
```

Comment:
There is also a question about self scope priority here.

Reply:
Max priority